### PR TITLE
Add ui for orchestrators, operators, and projects

### PR DESCRIPTION
**Pull request by abjjabjj**
https://www.loom.com/share/a3215351d1dd43d0afe8ee1ea84d89da?sid=a4962ad1-b9e7-4550-a55d-898119ede347

using `make` command does NOT provide multiple workers; can probably add options to `fastapi dev` but I personally used
```
uvicorn webapp.main.server:app --workers 3
```
in root directory to run with three workers for some concurrency.

Summary of changes:
Moved HTML outputs in `webapp/main/server.py` to Jinja templates as much as possible.

Large changes to our DB Models. Creating an operator automatically creates a DM project, basic user model added, messaging models updated. Project model hardcoded for software project for the sake of demo-ing.

Frontend has websocket to receive real time updates on compute, as well as DB message access for persistence.
You can create projects inside of orchestrators now and chat for those projects.

The loading SVG for prompt generation is broken due to the move to websocket; will be addressed in upcoming styling PR.

After doing some research, I'm also inclined to replace the websockets with SSE; I feel it more accurately reflects the interface since this is not a true chat experience, and it should allow us to provide a better UX a la loading animations and knowing when a prompt has actually finished executing.

### CODE REVIEW

File: concrete/abstract.py [hunk 1]
```diff
@@ -145,6 +160,19 @@ def _qna(
...
 
         answer = response.parsed
+
+        with Session() as session:
+            crud.create_message(
+                session,
+                MessageCreate(
+                    type_name=response_format.__name__,
+                    content=str(answer),
+                    prompt=self.prompt,
```
Comment (MichaelDeng03) [inline concrete/abstract.py:170]: what's the difference between self.prompt and query?

File: concrete/db/crud.py [hunk 2]
```diff
@@ -65,8 +68,26 @@ def delete_generic(db: Session, model: M | None) -> M | None:
 # ===Operator=== #
 
 
+# TODO: automate project creation via DML trigger/event
 def create_operator(db: Session, operator_create: OperatorCreate) -> Operator:
-    return create_generic(db, Operator(**operator_create.model_dump()))
+    project_create = ProjectCreate(
```
Comment (MichaelDeng03) [inline concrete/db/crud.py:73]: So create_operator can only be used for creating DM operators, and can't be used generically, e.g. adding operators to an orchestration?

File: concrete/db/orm/models.py [hunk 3]
```diff
@@ -16,6 +19,13 @@ class MetadataMixin(SQLModel):
     id: UUID = Field(primary_key=True, default_factory=uuid4)
 
 
+class ProfilePictureMixin(SQLModel):
```
Comment (MichaelDeng03) [inline concrete/db/orm/models.py:22]: ooh kinda neat

File: concrete/db/orm/models.py [hunk 4]
```diff
@@ -24,6 +34,25 @@ class OperatorToolLink(Base, table=True):
     tool_id: UUID = Field(foreign_key="tool.id", primary_key=True)
 
 
+# User Models
+
+
+class UserBase(Base, ProfilePictureMixin):
```
Comment (MichaelDeng03) [inline concrete/db/orm/models.py:40]: is this todo for after oauth?

File: concrete/db/orm/models.py [hunk 5]
```diff
@@ -88,7 +122,15 @@ class Operator(OperatorBase, MetadataMixin, table=True):
         cascade_delete=True,
     )
     tools: list["Tool"] = Relationship(back_populates="operators", link_model=OperatorToolLink)
-    orchestrator: "Orchestrator" = Relationship(back_populates="operators")
+    orchestrator: Orchestrator = Relationship(back_populates="operators")
```
Comment (MichaelDeng03) [inline concrete/db/orm/models.py:125]: this kinda melts my brain, dym adding some comments?

File: concrete/db/orm/models.py [hunk 6]
```diff
@@ -141,6 +183,71 @@ class Client(ClientBase, MetadataMixin, table=True):
...
+
+class ProjectBase(Base):
+    title: str = Field(description="Title of the project.", max_length=64)
+    orchestrator_id: UUID = Field(
+        description="ID of Orchestrator that owns this project.",
+        foreign_key="orchestrator.id",
+        ondelete="CASCADE",
+    )
+
+    executive_id: UUID | None = Field(
```
Comment (MichaelDeng03) [inline concrete/db/orm/models.py:197]: feels a little weird to hardcode executive/dev. does adding project_id as a relation to operators not work?

File: concrete/db/orm/models.py [hunk 7]
```diff
@@ -141,6 +183,71 @@ class Client(ClientBase, MetadataMixin, table=True):
...
+        sa_relationship_kwargs={
+            "foreign_keys": "Project.direct_message_operator_id",
+            "single_parent": True,
+        },
+    )
+
+    executive: Operator | None = Relationship(sa_relationship_kwargs={"foreign_keys": "Project.executive_id"})
+    developer: Operator | None = Relationship(sa_relationship_kwargs={"foreign_keys": "Project.developer_id"})
+
+    messages: list["Message"] = Relationship(back_populates="project", cascade_delete=True)
```
Comment (MichaelDeng03) [inline concrete/db/orm/models.py:248]: noice

File: concrete/db/orm/models.py [hunk 8]
```diff
@@ -173,12 +280,33 @@ class MessageBase(Base):
...
+        foreign_key="user.id",
+        default=None,
+    )
+    operator_id: UUID | None = Field(
+        description="ID of this message's sender if it's an operator.",
+        foreign_key="operator.id",
+        default=None,
+    )
+
+    @model_validator(mode="after")
```
Comment (MichaelDeng03) [inline concrete/db/orm/models.py:300]: 🤯

File: concrete/orchestrator.py [hunk 9]
```diff
@@ -179,26 +178,38 @@ class SoftwareOrchestrator(Orchestrator, StatefulMixin):
...
         self.state = State(self, orchestrator=self)
         self.uuid = uuid1()
         self.clients = {
             "openai": OpenAIClient(),
         }
-        self.operators: dict[str, AbstractOperator_co] = {
-            "exec": Executive(self.clients),
-            "dev": Developer(self.clients),
-        }
+        self.exec = Executive(self.clients, operator_id=executive_id)
```
Comment (MichaelDeng03) [inline concrete/orchestrator.py:191]: did you intend on updating abstractoperator with a sqlmodel getter?

File: webapp/main/server.py [hunk 10]
```diff
@@ -1,78 +1,79 @@
...
 templates = Jinja2Templates(directory=os.path.join(dname, "templates"))
+pages = Jinja2Templates(directory=os.path.join(dname, "templates", "pages"))
+components = Jinja2Templates(directory=os.path.join(dname, "templates", "components"))
 app.mount("/static", StaticFiles(directory=os.path.join(dname, "static")), name="static")
 
 
 annotatedFormStr = Annotated[str, Form()]
+annotatedFormUuid = Annotated[UUID, Form()]
+
+
```
Comment (MichaelDeng03) [inline webapp/main/server.py:36]: wooo

File: webapp/main/server.py [hunk 11]
```diff
@@ -1,78 +1,79 @@
...
-                    </a>
-                    <button hx-delete="/orchestrators/{o.id}" hx-swap="none">Delete</button>
-                </li>
-                """
-                for o in orchestrators
-            ]
+        CLIClient.emit_sequence(orchestrators)
+        CLIClient.emit("\n")
+        return components.TemplateResponse(
+            name="orchestrator_list.html",
```
Comment (MichaelDeng03) [inline webapp/main/server.py:74]: nice template

File: webapp/main/server.py [hunk 12]
```diff
@@ -181,51 +150,164 @@ async def create_operator(
     title: annotatedFormStr,
 ):
     # TODO: keep tabs on proper integration of Pydantic and Form. not working as expected from the FastAPI docs
```
Comment (MichaelDeng03) [inline webapp/main/server.py:152]: 👀

File: concrete/abstract.py (hunk 1 above)
Comment (abjjabjj) [reply concrete/abstract.py:170]: prompt is the starting prompt. I'll rename for clarity

File: concrete/db/crud.py (hunk 2 above)
Comment (abjjabjj) [reply concrete/db/crud.py:73]: The right way to think about it is that every operator gets its own DM. The workflow is also to create operators inside an orchestration.

File: concrete/db/orm/models.py (hunk 3 above)
Comment (abjjabjj) [reply concrete/db/orm/models.py:22]: probably will expand to other profile data as well

File: concrete/db/orm/models.py (hunk 4 above)
Comment (abjjabjj) [reply concrete/db/orm/models.py:40]: which TODO? the urllib one?

Comment (MichaelDeng03) [reply concrete/db/orm/models.py:40]: All the user models

Comment (abjjabjj) [reply concrete/db/orm/models.py:40]: oh yeah. it needs oauth 4sure

File: concrete/orchestrator.py (hunk 9 above)
Comment (abjjabjj) [reply concrete/orchestrator.py:191]: do you mean merging together the DB operator and compute operator? [](edited-with-Axolo)

File: concrete/db/orm/models.py (hunk 5 above)
Comment (abjjabjj) [reply concrete/db/orm/models.py:125]: are you talking about the `link_model`? [](edited-with-Axolo)

File: concrete/db/orm/models.py (hunk 6 above)
Comment (abjjabjj) [reply concrete/db/orm/models.py:197]: You could, but for this current iteration you'd still need to specify the role of the operator in the project. [](edited-with-Axolo)

File: concrete/orchestrator.py (hunk 9 above)
Comment (MichaelDeng03) [reply concrete/orchestrator.py]: Yeah, what you have seems fine though

File: webapp/main/server.py [hunk 13]
```diff
@@ -87,90 +98,58 @@ async def create_orchestrator(
...
 
 
 @app.delete("/orchestrators/{orchestrator_id}")
 async def delete_orchestrator(orchestrator_id: UUID):
     with Session() as session:
-        crud.delete_orchestrator(session, orchestrator_id)
+        orchestrator = crud.delete_orchestrator(session, orchestrator_id)
+        CLIClient.emit(f"{orchestrator}\n")
         headers = {"HX-Trigger": "getOrchestrators"}
         return HTMLResponse(content="", headers=headers)
```
Comment (MichaelDeng03) [inline webapp/main/server.py:137]:
```suggestion
        return HTMLResponse(content=f"Deleted orchestrator {orchestrator_id}", headers=headers)
```

**Review by MichaelDeng03 — APPROVED** [review]
lgtm, but your'e gonna have to create a new alembic migration - idk how to deal with migration conflicts, so probably wait until mr. lad merges his migration in

**Comment by abjjabjj** [comment]
ope @kgang

**Comment by kgang** [comment]
damn

**Comment by kgang** [comment]
I'm furiously adding env vars to the deploy scripts lmao

**Comment by kgang** [comment]
agh

**Comment by kgang** [comment]
we don't have an api deploy flow do we yet

**Comment by kgang** [comment]
I think we lose lmfao

**Comment by kgang** [comment]
:michil-rules:

**Comment by kgang** [comment]
:adequately-juiced:

**Pull request merged**